### PR TITLE
feat: add fetching user data

### PR DESCRIPTION
**Pull request by toshuyu**

### CODE REVIEW

**Comment by toshuyu** [comment]
Here's RSD review on behalf of masnec. (review count: 1)
 I have reviewed the provided Pull Request and found the following points:

1. There is a bug in the `getUserData` function. The function always returns `undefined` because `fetchUserData` is asynchronous and the `userData` variable is assigned before the callback function is executed. To fix this, you can either return a Promise from `getUserData` or use async/await syntax.
2. There are no unit tests provided in the diff. You should add unit tests to ensure that the functions are working as expected. Here's an example of how you can write unit tests using Jest:
```javascript
describe('fetchUserData', () => {
  it('should return user data when userId exists', (done) => {
    fetchUserData(1, (err, userData) => {
      expect(err).toBeNull();
      expect(userData).toEqual({ name: 'John Doe', age: 30 });
      done();
    });
  });

  it('should return error when userId does not exist', (done) => {
    fetchUserData(3, (err, userData) => {
      expect(err).toEqual(new Error('Usr not found'));
      expect(userData).toBeUndefined();
      done();
    });
  });
});

describe('getUserData', () => {
  it('should return user data when userId exists', async () => {
    const userData = await getUserData(1);
    expect(userData).toEqual({ name: 'John Doe', age: 30 });
  });

  it('should return error when userId does not exist', async () => {
    await expect(getUserData(3)).rejects.toThrow('Usr not found');
  });
});
```
3. Since a function related to User Data (`fetchUserData`) has been added, you should add API doc to the team's shared folder.

Here's the updated code with the suggested changes:
```javascript
function fetchUserData(userId, callback) {
  setTimeout(() => {
    const users = {
      1: { name: 'John Doe', age: 30 },
      2: { name: 'Jane Doe', age: 25 }
    };
    if (users[userId]) {
      callback(null, users[userId]); // First argument for error, second for succeess
    } else {
      callback(new Error('Usr not found'));
    }
  }, 1000); // Simulate netwok delay
}

function getUserData(userId) {
  return new Promise((resolve, reject) => {
    fetchUserData(userId, (err, userData) => {
      if (err) {
        reject(err);
      } else {
        resolve(userData);
      }
    });
  });
}

// Trying to use the function
getUserData(1)
  .then(userData => {
    console.log(userData);
  })
  .catch(err => {
    console.error(err);
  });
```

**Comment by toshuyu** [comment]
Here's RSD review on behalf of derwin162. (review count: 1)
 I have reviewed the provided Pull Request and will now address the tasks and code according to software engineering best practices, clean code, maintainability, performance, and security.

1. Typo and spelling check:
- In the last line of the code, there is a typo: 'userDeta' should be 'userData'.

2. Variable naming:
- The variable 'a' in the fetchUserData callback function is not a meaningful name. It should be renamed to something more descriptive, like 'user'.

3. Clean code and maintainability:
- The fetchUserData function is well-written and easy to understand.
- The getUserData function, however, is not returning the user data as expected due to the asynchronous nature of the callback. It is recommended to either use Promises or async/await to handle asynchronous operations more intuitively. Here is an example using Promises:

```javascript
function fetchUserData(userId) {
    return new Promise((resolve, reject) => {
        setTimeout(() => {
            const users = {
                1: { name: 'John Doe', age: 30 },
                2: { name: 'Jane Doe', age: 25 }
            };
            if (users[userId]) {
                resolve(users[userId]);
            } else {
                reject(new Error('User not found'));
            }
        }, 1000);
    });
}

async function getUserData(userId) {
    try {
        const userData = await fetchUserData(userId);
        return userData;
    } catch (error) {
        console.error(error);
    }
}

// Trying to use the function
getUserData(1)
    .then(userData => console.log(userData));
```

4. Performance:
- The performance of the code is not significantly impacted by the given example. However, in real-world applications, it is essential to minimize network requests and optimize database queries.

5. Security:
- The provided code does not contain any security vulnerabilities. However, in a real-world application, it is crucial to handle errors and edge cases more gracefully, validate user inputs, and sanitize data when interacting with databases or external APIs.

**Comment by toshuyu** [comment]
test doublecheck for masnec

**Comment by toshuyu** [comment]
test doublecheck for derwin162

**Comment by toshuyu** [comment]
Here's RSD review on behalf of masnec. (review count: 2)
 I have reviewed the provided Pull Request and will now address each task outlined by the engineer, as well as my own observations based on software engineering best practices.

1. Bugs:
- In the `getUserData` function, the `userData` variable may not be assigned the correct value due to the asynchronous nature of the `fetchUserData` function. This can lead to `undefined` being logged instead of the user data.

2. Unit tests:
- The Pull Request includes unit tests for both `fetchUserData` and `getUserData` functions, which is good. The tests cover positive and negative scenarios, ensuring that the functions behave as expected.

3. Adding API doc for user data functions:
- The `fetchUserData` function has been added, which handles user data. However, the provided diff does not contain any API documentation. Please make sure to add the necessary API documentation for this function in the team's shared folder.

4. Additional recommendations:
- In the `getUserData` function, consider using Promises instead of callbacks for better readability and easier error handling.
- In the unit tests for `getUserData`, you can improve the test for the case when the user is not found by checking if the Promise is rejected with the correct error message.

Here's an updated version of the code with the suggested changes:

```javascript
// script.js
function fetchUserData(userId) {
  return new Promise((resolve, reject) => {
    setTimeout(() => {
      const users = {
        1: { name: 'John Doe', age: 30 },
        2: { name: 'Jane Doe', age: 25 }
      };
      if (users[userId]) {
        resolve(users[userId]);
      } else {
        reject(new Error('Usr not found'));
      }
    }, 1000);
  });
}

async function getUserData(userId) {
  try {
    const userData = await fetchUserData(userId);
    return userData;
  } catch (error) {
    console.error(error);
  }
}

// Trying to use the function
getUserData(1).then(userData => console.log(userData));
```

```javascript
// test.js
describe('fetchUserData', () => {
  it('should return user data when userId exists', (done) => {
    fetchUserData(1).then(userData => {
      expect(userData).toEqual({ name: 'John Doe', age: 30 });
      done();
    });
  });

  it('should return error when userId does not exist', (done) => {
    fetchUserData(3).catch(err => {
      expect(err).toEqual(new Error('Usr not found'));
      done();
    });
  });
});

describe('getUserData', () => {
  it('should return user data when userId exists', async () => {
    const userData = await getUserData(1);
    expect(userData).toEqual({ name: 'John Doe', age: 30 });
  });

  it('should return error when userId does not exist', async () => {
    await expect(getUserData(3)).rejects.toThrow('Usr not found');
  });
});
```

**Comment by toshuyu** [comment]
Here's RSD review on behalf of derwin162. (review count: 2)
 I have reviewed the provided Pull Request and took into account the tasks outlined by the engineer and my knowledge of software engineering best practices. Here are my findings:

### Spelling and typos

The code is free of spelling and typos.

### Variable naming

The variable naming is meaningful and follows common conventions, making the code easy to understand.

### Code quality

1. In the `fetchUserData` function, it is a good practice to handle the error case first, before processing the success case. This way, the reader knows that error handling is taken care of, and the main logic flow is easier to follow. You can update the function as follows:

```javascript
function fetchUserData(userId, callback) {
    setTimeout(() => {
        const users = {
            1: { name: 'John Doe', age: 30 },
            2: { name: 'Jane Doe', age: 25 }
        };
        if (!users[userId]) {
            callback(new Error('Usr not found'));
            return;
        }
        callback(null, users[userId]); // First argument for error, second for success
    }, 1000); // Simulate network delay
}
```

2. In the `getUserData` function, the `userData` variable is not defined in the scope where it is returned. Instead, you can directly return the result of the callback:

```javascript
function getUserData(userId) {
    fetchUserData(userId, (error, user) => {
        if (error) {
            console.error(error);
            return;
        }
        return user;
    });
}
```

However, this implementation will still not work as expected because of JavaScript's asynchronous nature. You should either use promises or callbacks to handle the asynchronous behavior correctly. Here's an example using promises:

```javascript
function fetchUserData(userId) {
    return new Promise((resolve, reject) => {
        setTimeout(() => {
            const users = {
                1: { name: 'John Doe', age: 30 },
                2: { name: 'Jane Doe', age: 25 }
            };
            if (users[userId]) {
                resolve(users[userId]);
            } else {
                reject(new Error('Usr not found'));
            }
        }, 1000);
    });
}

async function getUserData(userId) {
    try {
        const user = await fetchUserData(userId);
        return user;
    } catch (error) {
        console.error(error);
    }
}
```

### Test quality

The test code looks good and covers the main use cases. However, the second test in the `getUserData` suite does not check if the error is thrown correctly. You can update it as follows:

```javascript
it('should return error when userId does not exist', async () => {
    await expect(getUserData(3)).rejects.toEqual(new Error('Usr not found'));
});
```

In conclusion, the code is mostly clean and follows good practices. However, there are some issues related to asynchronous behavior and error handling that need to be addressed. Additionally, the tests should be updated to ensure proper error handling.

**Comment by toshuyu** [comment]
test done for derwin162

**Comment by toshuyu** [comment]
Hi @derwin162,
RSD has helped you finished review. Please do manual review now, thanks.
